index.js

```diff
@@ -8,248 +8,260 @@ const {
-                }
-                return n
-              })
-            })
-          }
-
-          if (Tx.NFTokenOffers != null) {
-            _Tx.NFTokenOffers = Tx.NFTokenOffers
-          }
-
-          if (Tx.metaData != null) {
-            _Tx.Metadata = JSON.stringify(Tx.metaData)
-          }
-
-          CurrencyFields.forEach(CurrencyField => {
-            if (typeof Tx[CurrencyField] === 'string') {
-              Object.assign(_Tx, {
-                [CurrencyField + 'XRP']: parseInt(Tx[CurrencyField])
-              })
-            }
-            if (typeof Tx[CurrencyField] === 'object' && typeof Tx[CurrencyField].currency !== 'undefined') {
-              Object.assign(_Tx, {
-                [CurrencyField + 'DEX']: {
-                  currency: Tx[CurrencyField].currency,
-                  issuer: Tx[CurrencyField].issuer,
-                  value: parseFloat(Tx[CurrencyField].value)
-                }
-              })
-            }
-          })
-
-          // Special handling for timestamps
-          _Tx._InsertedAt = bigquery.timestamp(new Date())
-          
-          return _Tx
-        })
-        
-        // console.dir(Transactions[0], { depth: null })
-        // process.exit(1)
-
-        bigquery.dataset(DATASET_NAME).table(TRANSACTION_TABLE_NAME).insert(Transactions)
-          .then(r => {
-            console.log(`Inserted rows`, r)
-            LastLedger = Result.ledger_index
-            // process.exit(0)
-          })
-          .catch(err => {
-            if (err && err.name === 'PartialFailureError') {
-              if (err.errors && err.errors.length > 0) {
-                console.log('Insert errors:')
-                err.errors.forEach(err => console.dir(err, { depth: null }))
-                process.exit(1)
-              }
-            } else {
-              console.error('ERROR:', err)
-              process.exit(1)
-            }
-          })
+  return result[0][0].MaxLedger
+}
+
+async function fetchLedgerTransactions(client, ledgerIndex) {
+  const result = await client.send({
+    command: 'ledger',
+    ledger_index: ledgerIndex,
+    transactions: true,
+    expand: false,
+  }, xrplRequestOptions)
+
+  if (typeof result.ledger.transactions === 'undefined' || result.ledger.transactions.length === 0) {
+    return { ledger_index: ledgerIndex, transactions: [] }
+  }
+
+  if (result.ledger.transactions.length <= 200) {
+    const txResults = await client.send({
+      command: 'ledger',
+      ledger_index: ledgerIndex,
+      transactions: true,
+      expand: true,
+    }, xrplRequestOptions)
+    return { ledger_index: ledgerIndex, transactions: txResults.ledger.transactions }
+  }
+
+  console.log(`<<< MANY TXS at ledger ${ledgerIndex}: [[ ${result.ledger.transactions.length} ]], processing per-tx...`)
+  const txPromises = result.ledger.transactions.map((tx) => {
+    return client.send({
+      command: 'tx',
+      transaction: tx,
+    }, xrplRequestOptions)
+  })
```

Comment:
do you potentially run into a rate limit here if you're sending a ton of client requests at the same time?

Reply:
if you were using the promise-utils lib i linked you earlier, a mapLimit could be used to limit the number of parallel requests happening (https://blend.github.io/promise-utils/latest/index.html#maplimit)